File: python/src/runapi/grok_imagine/resources/segment_map.py
"""Grok-Imagine Image 2.0 segment-map resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import CompletedSegmentMapTaskResponse, SegmentMapTaskResponse


class SegmentMap(Resource):
    """Create an editable segmentation map from an Image 2.0 text-to-image task."""

    ENDPOINT = "/api/v1/grok_imagine/segment_map"

    RESPONSE_CLASS = SegmentMapTaskResponse
    COMPLETED_RESPONSE_CLASS = CompletedSegmentMapTaskResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["segment-map"], params)

File: ruby/lib/runapi/grok_imagine/resources/segment_map.rb
# frozen_string_literal: true

module RunApi
  module GrokImagine
    module Resources
      # Produces an editable segmentation map from a public image URL.
      # source_task_id remains accepted as a compatibility input; use image_url instead.
      class SegmentMap
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/grok_imagine/segment_map"

        RESPONSE_CLASS = Types::SegmentMapTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedSegmentMapTaskResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["segment-map"], params)
        end
      end
    end
  end
end
